### ISSUE TITLE
Last resilience4j release fully support Reactor , RxJava2 and completable future in their spring boot starters AOP

### ISSUE THREAD

**Issue by Romeh** [spring-cloud-incubator/spring-cloud-circuitbreaker#27]
Hey All ,

Small question .
Last resilience4j release fully support Reactor , RxJava2 and completable future in their spring boot starters AOP , can we reuse it here or the spring cloud starter here is serving different needs  like have unified way to define circuit breaker whatever the implementation behind it ?

i can help if you would like.

Thanks a lot,

**Comment by ryanjbaxter** [comment]
@Romeh can you describe more about how you would think it would change the implementation?  We already support reactive types

**Comment by Romeh** [comment]
@ryanjbaxter 

Right now the resilience4j spring support the following return types :

- Completablefutre
- RxJava2
- Reactor
- synchronous java types 

Plus the spring aspect annotation support (maybe it is not planned here for the spring cloud circuit breaker usage ).

can we add the missing 2 types as well (RxJava2 , completable futures)?

Thanks a lot,
Romeh

**Comment by Romeh** [comment]
@ryanjbaxter  gentle reminder for your valuable feedback ;)

**Comment by ryanjbaxter** [comment]
@Romeh thanks for the reminder.  The team was having a face to face meeting this week so I was waiting to talk about it in person.  What we were leaning towards possibly having a `CircuitBreaker` interface that returns a `Publisher` rather than `Mono` and `Flux`.  I think that would cover most of the cases above.

Monitoring is something we will want to look into as well, at least come up with examples showing that because the underlying library is going to expose that information for us.

**Comment by Romeh** [comment]
@ryanjbaxter ,

For the type of metrics and health endpoints exported under spring boot starter for resilience4j are documented here : 
https://resilience4j.readme.io/docs/getting-started-3#section-metrics-endpoint

- we have also micrometer exporter which can be used with dashboards: https://resilience4j.readme.io/docs/micrometer
- Grafana support : https://resilience4j.readme.io/docs/grafana-1 

Please ping me whenever u have questions !

**Comment by ryanjbaxter** [comment]
Regarding supporting other `Publisher` types.  It did not seem to make sense to try and support such a generic interface, as the `Publisher` interface itself does not offer much functionality and we found that the user would have to cast it back to the type they passed in anyways.  We could try and figure out the type but that might end badly if we guess wrong.  So in the end we decided to leave the interface as it and keep it to just `Mono` and `Flux`.  That said if users do want to use something like RxJava2 or Futures etc they can use the `RxJava2Adapter` from Project Reactor or the methods themselves on `Mono` and `Flux` to adapt to the APIs.  We can always add direct support for these libraries if we find a strong need to.
